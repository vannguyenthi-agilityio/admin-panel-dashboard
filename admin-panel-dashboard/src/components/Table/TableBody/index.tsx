import { memo } from "react";

// Types
import { IColumnType } from "@/types";

// Helpers
import { getObjectValue } from "@/helpers";

// Constants
import { RESULT_NOT_FOUND } from "@/constants";

interface IDataTableBody<T> {
  data: T[];
  columns: IColumnType<T>[];
}

const TableBody = <T,>({ data, columns }: IDataTableBody<T>) => {
  return (
    <tbody>
      {data.length ? (
        data.map((item, index )=> {
          const id = getObjectValue(item, "id");
          return (
            <tr key={`table-body-${id}`}>
              {columns.map(column => (
                <td
                  key={`table-row-cell-${column.key}`}
                  className={`px-6 py-2 border-0 border-b border-gray-100 ${index % 2 ? "bg-transparent" : "bg-greyish"}`}>
                  {column.customNode ? (
                    column.customNode(column, item)
                  ) : (
                    <div className="flex items-center">
                      {column.key === "status" && (getObjectValue(item, column.key) === "Approved" ?  <div className="h-2.5 w-2.5 rounded-full bg-green-500 me-2" /> :
                      getObjectValue(item, column.key) === "Rejected" ? <div className="h-2.5 w-2.5 rounded-full bg-red-500 me-2" /> : <div className="h-2.5 w-2.5 rounded-full bg-gray-500 me-2" />
                        )
                      }
                      <p className="text-darker font-normal text-sm">
                        {getObjectValue(item, column.key)}
                      </p>
                    </div>
                  )}
                </td>
              ))}
            </tr>
          );
        })
      ) : (
        <tr className="w-full">
          <td colSpan={10} className="h-32 text-center">
            <p className="text-xl font-semibold">
              {RESULT_NOT_FOUND}
            </p>
          </td>
        </tr>
      )}
    </tbody>
  );
};

export default memo(TableBody) as typeof TableBody;
