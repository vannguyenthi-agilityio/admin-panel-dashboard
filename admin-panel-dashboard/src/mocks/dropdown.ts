// Types
import { IListDropDown, IUser } from "@/types";

export const MOCK_USER_DATA: IUser = {
  name: "Derek Alvarado",
  avatar: "https://www.gravatar.com/avatar/2c7d99fe281ecd3bcd65ab915bac6dd5?s=250"
}

export const LIST_DROPDOWN_DATA: IListDropDown[] = [
  {
    href: "/",
    name: "Dashboard"
  },
  {
    href: "/",
    name: "Settings"
  }
]
