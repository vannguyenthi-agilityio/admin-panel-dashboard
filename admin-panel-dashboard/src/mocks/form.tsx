// Types
import { ICustomerData } from "@/types";

export const MOCK_CUSTOMER_DATA: ICustomerData = {
  id: 1,
  idNumber: 123,
  email: "test@gmail.com",
  firstName: "Alyvia",
  lastName: "Kelley",
  phoneNumber: "20 1234 4567",
  dateOfBirth: {months: '', date: 1, years: 1}
}

export const MOCK_INIT_CUSTOMER_DATA: ICustomerData = {
  id: "",
  idNumber: 0,
  email: "",
  firstName: "",
  lastName: "",
  phoneNumber: "",
  dateOfBirth: {months: '', date: 0, years: 0}
}
