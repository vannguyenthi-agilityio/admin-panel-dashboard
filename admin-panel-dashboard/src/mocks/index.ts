export * from './dropdown';
export * from './table';
export * from './form';
