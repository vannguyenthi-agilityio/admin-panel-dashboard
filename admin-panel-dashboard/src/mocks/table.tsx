// Types
import { IColumnType, ICustomerTable } from "@/types";

export const MOCK_CUSTOMER: ICustomerTable[] = [
  {
    id: 1,
    dateOfBirth: "06/18/1978",
    status: 'Approved',
    email: "test@gmail.com",
    fullName: "Alyvia Kelley",
  },
  {
    id: 2,
    dateOfBirth: "09/18/1983",
    status: 'Rejected',
    email: "jaiden.n@gmail.com",
    fullName: "Jaiden Nixon",
  },
  {
    id: 3,
    dateOfBirth: "10/18/1988",
    status: 'Blocked',
    email: "me@clayton.com",
    fullName: "Jaiden Nixon",
  }
];

export const MOCK_COLUMNS: IColumnType<ICustomerTable>[] = [
  {
    key: "id",
    title: "#",
  },
  {
    key: "fullName",
    title: "Full Name",
    isSortable: true,
  },
  {
    key: "status",
    title: "Status",
    isSortable: false,
  },
  {
    key: "email",
    title: "E-Mail",
    isSortable: true,
  },
  {
    key: "dateOfBirth",
    title: "Date of Birth",
    isSortable: true,
  },
];
