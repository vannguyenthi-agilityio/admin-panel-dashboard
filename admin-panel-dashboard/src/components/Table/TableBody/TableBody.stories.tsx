// Libs
import type { Meta, StoryObj } from "@storybook/react";

// Components
import TableBody from ".";

// Mocks
import { MOCK_CUSTOMER, MOCK_COLUMNS } from "@/mocks";

// Types
import { IColumnType } from "@/types";

const meta: Meta<typeof TableBody> = {
  title: "Components/Table/TableBody",
  tags: ["autodocs"],
  component: TableBody,
  argTypes: {
    data: { description: "data of body table" },
    columns: { description: "columns of table" },
  },
};

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    data: MOCK_CUSTOMER,
    columns: MOCK_COLUMNS as IColumnType<unknown>[],
  },
};

export const Empty: Story = {
  args: {
    data: [],
    columns: MOCK_COLUMNS as IColumnType<unknown>[],
  },
};
