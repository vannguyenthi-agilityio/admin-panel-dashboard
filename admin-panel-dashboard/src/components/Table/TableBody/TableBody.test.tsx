import { render } from "@testing-library/react";
import { BrowserRouter } from 'react-router-dom';

// Mocks
import { MOCK_CUSTOMER, MOCK_COLUMNS } from "@/mocks";

// Components
import TableBody from ".";

describe("TableBody", () => {
  const mockActionCutomer = jest.fn();
  const mockProps = {
    data: MOCK_CUSTOMER,
    columns: MOCK_COLUMNS,
    onActionCutomer: mockActionCutomer
  };

  it("Should matches snapshot", async () => {
    const { container } = render(
      <BrowserRouter>
        <table>
          <TableBody {...mockProps} />
        </table>
      </BrowserRouter>
    );

    expect(container).toMatchSnapshot();
  });
});
