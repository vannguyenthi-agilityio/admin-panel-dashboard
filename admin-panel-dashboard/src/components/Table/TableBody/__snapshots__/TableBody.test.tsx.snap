// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableBody Should matches snapshot 1`] = `
<div>
  <table>
    <tbody>
      <tr>
        <td
          class="px-6 py-2 border-0 border-b border-gray-100 bg-greyish"
        >
          <div
            class="flex items-center"
          >
            <p
              class="text-darker font-normal text-sm max-w-[200px] truncate"
            >
              1
            </p>
          </div>
        </td>
        <td
          class="px-6 py-2 border-0 border-b border-gray-100 bg-greyish"
        >
          <div
            class="flex items-center"
          >
            <p
              class="text-darker font-normal text-sm max-w-[200px] truncate"
            >
              Alyvia Kelley
            </p>
          </div>
        </td>
        <td
          class="px-6 py-2 border-0 border-b border-gray-100 bg-greyish"
        >
          <div
            class="flex items-center"
          >
            <div
              class="h-2.5 w-2.5 rounded-full bg-green-500 me-2"
            />
            <p
              class="text-darker font-normal text-sm max-w-[200px] truncate"
            >
              Approved
            </p>
          </div>
        </td>
        <td
          class="px-6 py-2 border-0 border-b border-gray-100 bg-greyish"
        >
          <div
            class="flex items-center"
          >
            <p
              class="text-darker font-normal text-sm max-w-[200px] truncate"
            >
              test@gmail.com
            </p>
          </div>
        </td>
        <td
          class="px-6 py-2 border-0 border-b border-gray-100 bg-greyish"
        >
          <div
            class="flex items-center"
          >
            <p
              class="text-darker font-normal text-sm max-w-[200px] truncate"
            >
              06/18/1978
            </p>
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="px-6 py-2 border-0 border-b border-gray-100 bg-transparent"
        >
          <div
            class="flex items-center"
          >
            <p
              class="text-darker font-normal text-sm max-w-[200px] truncate"
            >
              2
            </p>
          </div>
        </td>
        <td
          class="px-6 py-2 border-0 border-b border-gray-100 bg-transparent"
        >
          <div
            class="flex items-center"
          >
            <p
              class="text-darker font-normal text-sm max-w-[200px] truncate"
            >
              Jaiden Nixon
            </p>
          </div>
        </td>
        <td
          class="px-6 py-2 border-0 border-b border-gray-100 bg-transparent"
        >
          <div
            class="flex items-center"
          >
            <div
              class="h-2.5 w-2.5 rounded-full bg-gray-500 me-2"
            />
            <p
              class="text-darker font-normal text-sm max-w-[200px] truncate"
            >
              Rejected
            </p>
          </div>
        </td>
        <td
          class="px-6 py-2 border-0 border-b border-gray-100 bg-transparent"
        >
          <div
            class="flex items-center"
          >
            <p
              class="text-darker font-normal text-sm max-w-[200px] truncate"
            >
              jaiden.n@gmail.com
            </p>
          </div>
        </td>
        <td
          class="px-6 py-2 border-0 border-b border-gray-100 bg-transparent"
        >
          <div
            class="flex items-center"
          >
            <p
              class="text-darker font-normal text-sm max-w-[200px] truncate"
            >
              09/18/1983
            </p>
          </div>
        </td>
      </tr>
      <tr>
        <td
          class="px-6 py-2 border-0 border-b border-gray-100 bg-greyish"
        >
          <div
            class="flex items-center"
          >
            <p
              class="text-darker font-normal text-sm max-w-[200px] truncate"
            >
              3
            </p>
          </div>
        </td>
        <td
          class="px-6 py-2 border-0 border-b border-gray-100 bg-greyish"
        >
          <div
            class="flex items-center"
          >
            <p
              class="text-darker font-normal text-sm max-w-[200px] truncate"
            >
              Jaiden Nixon
            </p>
          </div>
        </td>
        <td
          class="px-6 py-2 border-0 border-b border-gray-100 bg-greyish"
        >
          <div
            class="flex items-center"
          >
            <div
              class="h-2.5 w-2.5 rounded-full bg-red-500 me-2"
            />
            <p
              class="text-darker font-normal text-sm max-w-[200px] truncate"
            >
              Blocked
            </p>
          </div>
        </td>
        <td
          class="px-6 py-2 border-0 border-b border-gray-100 bg-greyish"
        >
          <div
            class="flex items-center"
          >
            <p
              class="text-darker font-normal text-sm max-w-[200px] truncate"
            >
              me@clayton.com
            </p>
          </div>
        </td>
        <td
          class="px-6 py-2 border-0 border-b border-gray-100 bg-greyish"
        >
          <div
            class="flex items-center"
          >
            <p
              class="text-darker font-normal text-sm max-w-[200px] truncate"
            >
              10/18/1988
            </p>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
